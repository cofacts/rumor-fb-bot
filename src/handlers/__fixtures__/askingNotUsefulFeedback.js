export const oneFeedback = {
  data: {
    action: {
      feedbackCount: 1,
    },
  },
};

export const twoFeedbacks = {
  data: {
    action: {
      feedbackCount: 2,
    },
  },
};
