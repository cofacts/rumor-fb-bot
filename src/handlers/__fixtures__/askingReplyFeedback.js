export const twoFeedbacks = {
  data: {
    action: {
      feedbackCount: 2,
    },
    GetReply: {
      type: 'NOT_ARTICLE',
      text:
        '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
    },
    GetArticle: {
      text:
        '(0)(1)(/)(0)(9)(line)免費貼圖\n\n「[全螢幕貼圖]生活市集x生活小黑熊」\nhttps://line.me/S/sticker/10299\n\n「松果購物 x 狗與鹿(動態貼圖) 」\nhttps://line.me/S/sticker/10300\n\n「LINE TV X 我的男孩限免貼圖」\nhttps://line.me/S/sticker/10207',
    },
  },
};

export const oneFeedback = {
  data: {
    action: {
      feedbackCount: 1,
    },
  },
};

export const articleData = {
  data: {
    GetReply: {
      type: 'NOT_ARTICLE',
      text:
        '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
    },
    GetArticle: {
      text:
        '(0)(1)(/)(0)(9)(line)免費貼圖\n\n「[全螢幕貼圖]生活市集x生活小黑熊」\nhttps://line.me/S/sticker/10299\n\n「松果購物 x 狗與鹿(動態貼圖) 」\nhttps://line.me/S/sticker/10300\n\n「LINE TV X 我的男孩限免貼圖」\nhttps://line.me/S/sticker/10207',
    },
  },
};
