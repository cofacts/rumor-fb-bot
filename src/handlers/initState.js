import stringSimilarity from 'string-similarity';
import gql from '../gql';
import {
  createPostbackAction,
  isNonsenseText,
  createAskArticleSubmissionReply,
} from './utils';
import ga from '../ga';

const SIMILARITY_THRESHOLD = 0.95;

/**
 * The entry state for both messages and comments
 */
export default async function initState(params) {
  let { data, state, event, issuedAt, userId, replies, isSkipUser } = params;

  const visitor = ga('FB-' + userId, event.input);
  visitor.screenview({ screenName: state });
  visitor.event({ ec: 'UserInput', ea: 'MessageType', el: 'text' });

  // Track text message type send by user
  visitor.event({ ec: 'UserInput', ea: 'MessageType', el: 'text' });

  // Store user input into context
  data.searchedText = event.input;

  // Search for articles
  const {
    data: { ListArticles },
  } = await gql`
    query($text: String!) {
      ListArticles(
        filter: { moreLikeThis: { like: $text } }
        orderBy: [{ _score: DESC }]
        first: 4
      ) {
        edges {
          node {
            text
            id
          }
        }
      }
    }
  `({
    text: event.input,
  });

  const articleSummary = `${event.input.slice(0, 10)}${
    event.input.length > 10 ? '⋯⋯' : ''
  }`;

  if (ListArticles.edges.length) {
    // Track if find similar Articles in DB.
    visitor.event({ ec: 'UserInput', ea: 'ArticleSearch', el: 'ArticleFound' });
    // Track which Article is searched. And set tracking event as non-interactionHit.
    ListArticles.edges.forEach(edge => {
      visitor.event({
        ec: 'Article',
        ea: 'Search',
        el: edge.node.id,
        ni: true,
      });
    });

    const edgesSortedWithSimilarity = ListArticles.edges
      .map(edge => {
        edge.similarity = stringSimilarity.compareTwoStrings(
          // Remove spaces so that we count word's similarities only
          edge.node.text.replace(/\s/g, ''),
          event.input.replace(/\s/g, '')
        );
        return edge;
      })
      .sort((edge1, edge2) => edge2.similarity - edge1.similarity);

    // Store article ids
    data.foundArticleIds = edgesSortedWithSimilarity.map(
      ({ node: { id } }) => id
    );

    const hasIdenticalDocs =
      edgesSortedWithSimilarity[0].similarity >= SIMILARITY_THRESHOLD;

    if (userId === '0') {
      // from facebook comment

      const links = edgesSortedWithSimilarity.map(
        ({ node: { id } }) => `https://cofacts.g0v.tw/article/${id}`
      );

      // search the top article
      const {
        data: { GetArticle },
      } = await gql`
        query($id: String!) {
          GetArticle(id: $id) {
            replyCount
            articleReplies(status: NORMAL) {
              reply {
                id
                type
                text
              }
              positiveFeedbackCount
              negativeFeedbackCount
            }
          }
        }
      `({
        id: edgesSortedWithSimilarity[0].node.id,
      });

      const count = {};

      GetArticle.articleReplies.forEach(ar => {
        // Track which Reply is searched. And set tracking event as non-interactionHit.
        visitor.event({ ec: 'Reply', ea: 'Search', el: ar.reply.id }, true);

        const type = ar.reply.type;
        if (!count[type]) {
          count[type] = 1;
        } else {
          count[type]++;
        }
      });

      let summary =
        // and
        '，而且有：\n' +
        // {} person(s) consider this to be a rumor
        `${count.RUMOR ? `${count.RUMOR} 個人覺得 ❌ 含有不實訊息\n` : ''}` +
        // {} person(s) think this can be a truth
        `${
          count.NOT_RUMOR ? `${count.NOT_RUMOR} 個人覺得 ⭕ 含有真實訊息\n` : ''
        }` +
        // {} person(s) think this is simply personal opinion
        `${
          count.OPINIONATED
            ? `${count.OPINIONATED} 個人覺得 💬 含有個人意見\n`
            : ''
        }`;
      if (count.NOT_ARTICLE) {
        // but also {} person(s) thinks Cofacts need not to handle this message
        summary += `，不過有 ${
          count.NOT_ARTICLE
        } 個人覺得 ⚠️️ 不在 Cofacts查證範圍\n`;
      }

      const replies = [
        {
          type: 'text',
          content: {
            // Hey #Cofacts has messages {}% similar to this one! {summary}
            // Go to Cofacts' website for more information!
            // {Links}
            text: `#Cofacts 上有訊息跟這則有 ${Math.round(
              edgesSortedWithSimilarity[0].similarity * 100
            )}% 像${summary}\n到 Cofacts 上面看看相關訊息吧！\n${links.join(
              '\n'
            )}`,
          },
        },
      ];

      visitor.send();
      return {
        data,
        state,
        event,
        issuedAt,
        userId,
        replies,
        isSkipUser: false,
      };
    }

    // message
    if (edgesSortedWithSimilarity.length === 1 && hasIdenticalDocs) {
      // choose for user
      event.input = 1;

      visitor.send();
      return {
        data,
        state: 'CHOOSING_ARTICLE',
        event,
        issuedAt,
        userId,
        replies,
        isSkipUser: true,
      };
    }

    const templateMessage = {
      type: 'template',
      payload: {
        template_type: 'generic',
        elements: edgesSortedWithSimilarity
          .map(({ node: { text }, similarity }, idx) => ({
            title: text.slice(0, 80),
            // [Similarity: {}%]
            subtitle: `[相似度:${(similarity * 100).toFixed(2) + '%'}]`,
            // Choose this one
            buttons: [createPostbackAction('選擇此則', idx + 1)],
          }))
          .concat(
            hasIdenticalDocs
              ? []
              : [
                  {
                    // These messages don't match mine :(
                    title: '這裡沒有一篇是我傳的訊息。',
                    // Choose this one
                    buttons: [createPostbackAction('選擇', 0)],
                  },
                ]
          ),
      },
    };

    replies = [
      {
        type: 'text',
        content: {
          // We're checking "{articleSummary}" for you...
          text: `幫您查詢「${articleSummary}」的相關回應。`,
        },
      },
      {
        type: 'text',
        content: {
          // Which message below matches what you just sent to us?
          text: '請問下列文章中，哪一篇是您剛才傳送的訊息呢？',
        },
      },
      {
        type: 'carousel',
        content: {
          attachment: templateMessage,
          quick_replies: [
            {
              content_type: 'text',
              // I want to contact Cofacts team
              title: '我是要找 Cofacts 的人啦',
              // I want to contact Cofacts team
              payload: '我是要找 Cofacts 的人啦',
            },
          ],
        },
      },
    ];
    state = 'CHOOSING_ARTICLE';
  } else {
    if (isNonsenseText(event.input)) {
      // Track if find similar Articles in DB.
      visitor.event({
        ec: 'UserInput',
        ea: 'ArticleSearch',
        el: 'NonsenseText',
      });

      replies = [
        {
          type: 'text',
          content: {
            text:
              // Sorry, please provide more information.
              // Please refer to our user's manual http://bit.ly/cofacts-fb-users
              '你傳的資訊資訊太少，無法為你搜尋資料庫噢！\n' +
              '正確使用方式，請參考📖使用手冊 http://bit.ly/cofacts-fb-users',
          },
        },
      ];
      state = '__INIT__';
    } else {
      // Track if find similar Articles in DB.
      visitor.event({
        ec: 'UserInput',
        ea: 'ArticleSearch',
        el: 'ArticleNotFound',
      });

      if (userId === '0') {
        // from a comment
        // for issue #2, keep links in replies when nothing is found
        // since it contains less information that should be broadcast
        replies = [
          {
            type: 'text',
            content: {
              // We didn't find anything about {articleSummary} :(
              // You can try these websites again: ...
              // Or report this article to us!
              text: `找不到關於「${articleSummary}」的訊息耶 QQ\n可以嘗試到這些地方找找相關訊息：\n
              蘭姆酒吐司Rumor & Truth https://www.facebook.com/rumtoast/\n或者到 LINE 上面把謠言傳給我們~\nhttp://bit.ly/cofacts-line-users`,
            },
          },
        ];
      } else {
        replies = [
          {
            type: 'text',
            content: {
              // We didn't find anything about {articleSummary} :(
              text: `找不到關於「${articleSummary}」訊息耶 QQ`,
            },
          },
        ].concat(createAskArticleSubmissionReply());
        state = 'ASKING_ARTICLE_SUBMISSION_REASON';
      }
    }
  }

  visitor.send();
  return { data, state, event, issuedAt, userId, replies, isSkipUser };
}
