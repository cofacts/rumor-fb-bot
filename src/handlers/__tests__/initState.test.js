jest.mock('../../gql');

import initState from '../initState';
import * as apiResult from '../__fixtures__/initState';
import gql from '../../gql';

it('article found', async () => {
  gql.__push(apiResult.longArticle);

  const input = {
    data: {},
    state: '__INIT__',
    event: {
      type: 'message',
      input:
        '計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入',
      timestamp: 1497994016356,
      message: {
        type: 'text',
        id: '6270464463537',
        text:
          '計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入',
      },
    },
    issuedAt: 1497994017447,
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  expect(await initState(input)).toMatchSnapshot();
});

it('article found with 120 words limit', async () => {
  gql.__push(apiResult.twoLongArticles);

  const input = {
    data: {},
    state: '__INIT__',
    event: {
      type: 'message',
      input:
        '這樣的大事國內媒體竟然不敢報導！\n我國駐日代表將原「中華民國」申請更名為「台灣」結果被日本裁罰，須繳納7000萬日圓（合約台幣2100萬元）高額稅賦(轉載中時電子報）\n\n我駐日代表謝長廷將原「中華民國」申請更名為「台灣」，自認得意之時，結果遭自認友好日本國給出賣了，必須繳納7000萬日圓（合約台幣2100萬元）高額稅賦...民進黨沒想到如此更名竟然是這樣的下場：被他最信任也最友好的日本政府給坑了。\n果然錯誤的政策比貪污可怕，2100萬就這樣打水漂了，還要資助九州水患，核四停建違約賠償金.......夠全國軍公教退休2次.........\n\nhttp://www.chinatimes.com/newspapers/20170617000318-260118',
      timestamp: 1502477505309,
      message: {
        type: 'text',
        id: '6530038889933',
        text:
          '這樣的大事國內媒體竟然不敢報導！\n我國駐日代表將原「中華民國」申請更名為「台灣」結果被日本裁罰，須繳納7000萬日圓（合約台幣2100萬元）高額稅賦(轉載中時電子報）\n\n我駐日代表謝長廷將原「中華民國」申請更名為「台灣」，自認得意之時，結果遭自認友好日本國給出賣了，必須繳納7000萬日圓（合約台幣2100萬元）高額稅賦...民進黨沒想到如此更名竟然是這樣的下場：被他最信任也最友好的日本政府給坑了。\n果然錯誤的政策比貪污可怕，2100萬就這樣打水漂了，還要資助九州水患，核四停建違約賠償金.......夠全國軍公教退休2次.........\n\nhttp://www.chinatimes.com/newspapers/20170617000318-260118',
      },
    },
    issuedAt: 1502477506267,
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    isSkipUser: false,
  };

  const result = await initState(input);

  for (let reply of result.replies) {
    if (reply.type === 'text') {
      expect(reply.content.text.length < 2000).toBe(true);
    } else if (reply.type === 'buttons') {
      expect(reply.content.attachment.payload.text.length < 640).toBe(true);
    } else if (reply.type === 'carousel') {
      for (let element of reply.content.attachment.payload.elements) {
        expect(element.title.length <= 80).toBe(true);
        expect(element.subtitle.length <= 80).toBe(true);
      }
    }
  }
});

it('articles found with high similarity', async () => {
  gql.__push(apiResult.twoShortArticles);

  const input = {
    data: {},
    state: '__INIT__',
    event: {
      type: 'message',
      input: 'YouTube · 寻找健康人生',
      timestamp: 1497994016356,
      message: {
        type: 'text',
        id: '6270464463537',
        text: 'YouTube · 寻找健康人生',
      },
    },
    issuedAt: 1497994017447,
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  expect(await initState(input)).toMatchSnapshot();
});

it('only one article found with high similarity', async () => {
  gql.__push(apiResult.shortArticle);

  const input = {
    data: {},
    state: '__INIT__',
    event: {
      type: 'message',
      input: 'YouTube · 寻找健康人生',
      timestamp: 1497994016356,
      message: {
        type: 'text',
        id: '6270464463537',
        text: 'YouTube · 寻找健康人生',
      },
    },
    issuedAt: 1497994017447,
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  expect(await initState(input)).toMatchSnapshot();
});

it('should handle article not found', async () => {
  gql.__push(apiResult.notFound);

  const input = {
    data: {},
    state: '__INIT__',
    event: {
      type: 'message',
      input:
        'YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...',
      timestamp: 1497994016356,
      message: {
        type: 'text',
        id: '6270464463537',
        text:
          'YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...',
      },
    },
    issuedAt: 1497994017447,
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  expect(await initState(input)).toMatchSnapshot();
});

afterEach(() => {
  gql.__reset();
});
