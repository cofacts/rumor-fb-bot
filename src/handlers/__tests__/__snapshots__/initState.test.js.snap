// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`article found 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVvY-yizyCdS-nWhuYWx",
    ],
    "searchedText": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
  },
  "event": Object {
    "input": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
    "message": Object {
      "id": "6270464463537",
      "text": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1497994017447,
  "replies": Array [
    Object {
      "content": Object {
        "text": "幫您查詢「計程車上有裝悠遊卡感⋯⋯」的相關回應。",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "text": "請問下列文章中，哪一篇是您剛才傳送的訊息呢？",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "attachment": Object {
          "payload": Object {
            "elements": Array [
              Object {
                "buttons": Array [
                  Object {
                    "payload": 1,
                    "title": "選擇此則",
                    "type": "postback",
                  },
                ],
                "subtitle": "[相似度:52.41%]",
                "title": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入，下車時使用老人悠遊卡，跳錶車資105元，優惠32元，只扣73元，哈哈，這是屬於我",
              },
              Object {
                "buttons": Array [
                  Object {
                    "payload": 0,
                    "title": "選擇",
                    "type": "postback",
                  },
                ],
                "title": "這裡沒有一篇是我傳的訊息。",
              },
            ],
            "template_type": "generic",
          },
          "type": "template",
        },
        "quick_replies": Array [
          Object {
            "content_type": "text",
            "payload": "我是要找 Cofacts 的人啦",
            "title": "我是要找 Cofacts 的人啦",
          },
        ],
      },
      "type": "carousel",
    },
  ],
  "state": "CHOOSING_ARTICLE",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`article found for comments 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVvY-yizyCdS-nWhuYWx",
    ],
    "searchedText": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
  },
  "event": Object {
    "input": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
    "message": Object {
      "id": "6270464463537",
      "text": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1497994017447,
  "replies": Array [
    Object {
      "content": Object {
        "text": "#Cofacts 上有訊息跟這則有 52% 像，而且有：
1 個人覺得 ❌ 含有不實訊息
1 個人覺得 ⭕ 含有真實訊息
1 個人覺得 💬 含有個人意見
，不過有 2 個人覺得 ⚠️️ 不在 Cofacts查證範圍

到 Cofacts 上面看看相關訊息吧！
https://cofacts.g0v.tw/article/AVvY-yizyCdS-nWhuYWx",
      },
      "type": "text",
    },
  ],
  "state": "__INIT__",
  "userId": "0",
}
`;

exports[`articles found with high similarity 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVvY-yizyCdS-nWhuYWx",
      "AVvY-yizyCdS-nWhuYWy",
    ],
    "searchedText": "YouTube · 寻找健康人生",
  },
  "event": Object {
    "input": "YouTube · 寻找健康人生",
    "message": Object {
      "id": "6270464463537",
      "text": "YouTube · 寻找健康人生",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1497994017447,
  "replies": Array [
    Object {
      "content": Object {
        "text": "幫您查詢「YouTube · ⋯⋯」的相關回應。",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "text": "請問下列文章中，哪一篇是您剛才傳送的訊息呢？",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "attachment": Object {
          "payload": Object {
            "elements": Array [
              Object {
                "buttons": Array [
                  Object {
                    "payload": 1,
                    "title": "選擇此則",
                    "type": "postback",
                  },
                ],
                "subtitle": "[相似度:100.00%]",
                "title": "YouTube  ·  寻找健康人生",
              },
              Object {
                "buttons": Array [
                  Object {
                    "payload": 2,
                    "title": "選擇此則",
                    "type": "postback",
                  },
                ],
                "subtitle": "[相似度:100.00%]",
                "title": "YouTube
·
寻找健康人生",
              },
            ],
            "template_type": "generic",
          },
          "type": "template",
        },
        "quick_replies": Array [
          Object {
            "content_type": "text",
            "payload": "我是要找 Cofacts 的人啦",
            "title": "我是要找 Cofacts 的人啦",
          },
        ],
      },
      "type": "carousel",
    },
  ],
  "state": "CHOOSING_ARTICLE",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`only one article found with high similarity 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVvY-yizyCdS-nWhuYWx",
    ],
    "searchedText": "YouTube · 寻找健康人生",
  },
  "event": Object {
    "input": 1,
    "message": Object {
      "id": "6270464463537",
      "text": "YouTube · 寻找健康人生",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": true,
  "issuedAt": 1497994017447,
  "replies": undefined,
  "state": "CHOOSING_ARTICLE",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`should handle article not found 1`] = `
Object {
  "data": Object {
    "searchedText": "YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...",
  },
  "event": Object {
    "input": "YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...",
    "message": Object {
      "id": "6270464463537",
      "text": "YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1497994017447,
  "replies": Array [
    Object {
      "content": Object {
        "text": "找不到關於「YouTube · ⋯⋯」訊息耶 QQ",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "text": "【送出訊息到公開資料庫？】
若這是「轉傳訊息」，而且您覺得這很可能是一則「謠言」，請將這則訊息送進公開資料庫建檔，讓好心人查證與回覆。

雖然您不會立刻收到查證結果，但可以幫助到未來同樣收到這份訊息的人。",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "text": "請把「為何您會覺得這是一則謠言」的理由傳給我們，幫助闢謠編輯釐清您有疑惑之處。",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "attachment": Object {
          "payload": Object {
            "buttons": Array [
              Object {
                "payload": "n",
                "title": "放棄送出",
                "type": "postback",
              },
            ],
            "template_type": "button",
            "text": "若要放棄，請按「放棄送出」。",
          },
          "type": "template",
        },
      },
      "type": "buttons",
    },
  ],
  "state": "ASKING_ARTICLE_SUBMISSION_REASON",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`should handle article not found for comments 1`] = `
Object {
  "data": Object {
    "searchedText": "YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...",
  },
  "event": Object {
    "input": "YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...",
    "message": Object {
      "id": "6270464463537",
      "text": "YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1497994017447,
  "replies": Array [
    Object {
      "content": Object {
        "text": "找不到關於「YouTube · ⋯⋯」的訊息耶 QQ
可以嘗試到這些地方找找相關訊息：

              蘭姆酒吐司Rumor & Truth https://www.facebook.com/rumtoast/
或者到 LINE 上面把謠言傳給我們~
http://bit.ly/cofacts-line-users",
      },
      "type": "text",
    },
  ],
  "state": "__INIT__",
  "userId": "0",
}
`;
